### PR TITLE
Issue #768

### PR DESCRIPTION
**Pull request by vibsid0986-zz**
Added a plugin with "com.act21.mamba.layout.alg.tree" which will create an indented tree layout for the ElkGraph with the corresponding requirements as mentioned in the issue description.

### CODE REVIEW

**Comment by soerendomroes** [comment]
@uruuru @le-cds Do we have a policy on naming conventions for new algorithms and how do we usually integrate them?

File: plugins/com.act21.mamba.layout.alg.tree/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/target/
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/.gitignore:1]: Usually, we have no separate gitignore for plugins. Why do you need it?

File: plugins/com.act21.mamba.layout.alg.tree/.settings/org.eclipse.jdt.core.prefs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+eclipse.preferences.version=1
+org.eclipse.jdt.core.compiler.codegen.inlineJsrBytecode=enabled
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=11
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/.settings/org.eclipse.jdt.core.prefs:3]: Compiler compliance is currently 1.8 (I know it also works for 11) but please do not change such things. I suggest a separate PR for this.

File: plugins/com.act21.mamba.layout.alg.tree/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+Bundle-Version: 1.0.0.qualifier
+Bundle-Vendor: ACT21
+Require-Bundle: org.eclipse.elk.core,
+ org.eclipse.elk.graph,
+ org.eclipse.elk.alg.common,
+ com.google.guava
+Automatic-Module-Name: com.act21.mamba.layout.alg.grid
+Export-Package: com.act21.mamba.layout.alg.tree,
+ com.act21.mamba.layout.alg.tree.options
+Bundle-RequiredExecutionEnvironment: JavaSE-11
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/META-INF/MANIFEST.MF:14]: JavaSE-1.8

File: plugins/com.act21.mamba.layout.alg.tree/build.properties [hunk 4]
```diff
@@ -0,0 +1,5 @@
+source.. = src/,\
+           src-gen/
+output.. = target/classes/
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/build.properties:3]: output.. = bin/

File: plugins/com.act21.mamba.layout.alg.tree/build.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+bin.includes = META-INF/,\
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/build.properties:4]: bin.includes = META-INF/,\
               .,\
               about.html,\
src.includes = about.html

File: plugins/com.act21.mamba.layout.alg.tree/src-gen/com/act21/mamba/layout/alg/tree/.gitignore [hunk 6]
```diff
@@ -0,0 +1,4 @@
+/.GridMetadataProvider.java._trace
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src-gen/com/act21/mamba/layout/alg/tree/.gitignore:1]: Separate gitignore should not be needed

File: plugins/com.act21.mamba.layout.alg.tree/src-gen/com/act21/mamba/layout/alg/tree/options/.gitignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+/.GridOptions.java._trace
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src-gen/com/act21/mamba/layout/alg/tree/options/.gitignore:1]: Separate gitignore should not be needed

File: plugins/com.act21.mamba.layout.alg.tree/src/META-INF/services/ILayoutMetaDataProvider [hunk 8]
```diff
@@ -0,0 +1 @@
+com.act21.mamba.layout.alg.tree.GridMetadataProvider
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src/META-INF/services/ILayoutMetaDataProvider:1]: I expected the MetaDataProvider to be in an options package.

File: plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/Tree.melk [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package com.act21.mamba.layout.alg.tree
+
+import com.act21.mamba.layout.alg.tree.TreeLayoutProvider
+import org.eclipse.elk.core.math.ElkPadding
+
+bundle { 
+    metadataClass TreeMetadataProvider
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/Tree.melk:7]: bundle {
    metadataClass **options**.TreeMetadataProvider

File: plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/Tree.melk [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    label "Num Columns"
+    description
+        "Number of columns."
+    default = 1
+    targets parents 
+}
+
+algorithm Tree(TreeLayoutProvider) {
+    label "Tree"
+    description "Please insert a short but informative description here"
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/Tree.melk:29]: Please provide a description

File: plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/Tree.melk [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+option numColumns : int {
+    label "Num Columns"
+    description
+        "Number of columns."
+    default = 1
+    targets parents 
+}
+
+algorithm Tree(TreeLayoutProvider) {
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/Tree.melk:27]: Usually, the algorithm is specified at the top and the algorithms below. I suggest to do this here too.

File: plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/TreeLayoutProvider.java [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+import org.eclipse.emf.common.util.EList;
+
+/**
+ * A simple layout algorithm class. This algorithm already supports a number of
+ * layout options, places nodes, and routes edges.....
+ */
+public class TreeLayoutProvider extends AbstractLayoutProvider {
+
+	@Override
+	public void layout(ElkNode layoutGraph, IElkProgressMonitor progressMonitor) {
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/TreeLayoutProvider.java:21]: You might want to consider t use the progressMonitor to log graphs after certain parts of you algorithm

File: plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/TreeLayoutProvider.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+		if (root != null) {
+			root.setDepth(depth);
+			nodeList.add(root);
+			for (ElkTreeNode child : root.getChildren()) {
+				createTreeNodeList(child, nodeList, depth + 1);
+			}
+		}
+	}
+
+	// Function to create tree of ElkTreeNode using root and edgesMap
```
Comment (soerendomroes) [inline plugins/com.act21.mamba.layout.alg.tree/src/com/act21/mamba/layout/alg/tree/TreeLayoutProvider.java:154]: I suggest to add javadoc instead of comments.

File: plugins/org.eclipse.elk.alg.common/.classpath [hunk 14]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.common/.classpath:3]: Again I think a migration to java 11 should be a separate PR and this file should not be affected by introducing a new algorithm.

File: plugins/org.eclipse.elk.alg.common/.settings/org.eclipse.jdt.core.prefs [hunk 15]
```diff
@@ -1,34 +1,41 @@
 eclipse.preferences.version=1
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.common/.settings/org.eclipse.jdt.core.prefs:1]: This file should not change

File: plugins/org.eclipse.elk.alg.disco/.classpath [hunk 16]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.disco/.classpath:1]: This file should not change

File: plugins/org.eclipse.elk.alg.disco/.settings/org.eclipse.jdt.core.prefs [hunk 17]
```diff
@@ -1,34 +1,41 @@
 eclipse.preferences.version=1
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.disco/.settings/org.eclipse.jdt.core.prefs:1]: This file should not change

File: plugins/org.eclipse.elk.alg.graphviz.dot/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 18]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.graphviz.dot/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: This file should not change

File: plugins/org.eclipse.elk.core.debug.grandom.ide/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 19]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.debug.grandom.ide/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.debug.grandom.ui/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 20]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.debug.grandom.ui/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.debug.grandom/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 21]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.debug.grandom/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.debug/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 22]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.debug/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.meta/.classpath [hunk 23]
```diff
@@ -3,7 +3,11 @@
 	<classpathentry kind="src" path="src"/>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.meta/.classpath:3]: Should not change

File: plugins/org.eclipse.elk.core.meta/.settings/org.eclipse.jdt.core.prefs [hunk 24]
```diff
@@ -1,34 +1,41 @@
 eclipse.preferences.version=1
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.meta/.settings/org.eclipse.jdt.core.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.meta/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 25]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.meta/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.service/.classpath [hunk 26]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.service/.classpath:1]: Should not change

File: plugins/org.eclipse.elk.core.service/.settings/org.eclipse.jdt.core.prefs [hunk 27]
```diff
@@ -1,34 +1,41 @@
 eclipse.preferences.version=1
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.service/.settings/org.eclipse.jdt.core.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core.ui/.classpath [hunk 28]
```diff
@@ -1,7 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.ui/.classpath:1]: Should not change

File: plugins/org.eclipse.elk.core.ui/.settings/org.eclipse.jdt.core.prefs [hunk 29]
```diff
@@ -1,34 +1,41 @@
 eclipse.preferences.version=1
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core.ui/.settings/org.eclipse.jdt.core.prefs:1]: Should not change

File: plugins/org.eclipse.elk.core/.classpath [hunk 30]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core/.classpath:1]: Should not change

File: plugins/org.eclipse.elk.core/.settings/org.eclipse.jdt.core.prefs [hunk 31]
```diff
@@ -1,34 +1,41 @@
 eclipse.preferences.version=1
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.core/.settings/org.eclipse.jdt.core.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.json.text.ide/.project [hunk 32]
```diff
@@ -5,11 +5,6 @@
 	<projects>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json.text.ide/.project:5]: Should not change

File: plugins/org.eclipse.elk.graph.json.text.ide/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 33]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json.text.ide/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.json.text.ui/.project [hunk 34]
```diff
@@ -5,11 +5,6 @@
 	<projects>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json.text.ui/.project:5]: Should not change

File: plugins/org.eclipse.elk.graph.json.text.ui/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 35]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json.text.ui/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.json.text/.project [hunk 36]
```diff
@@ -5,11 +5,6 @@
 	<projects>
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json.text/.project:5]: Should not change

File: plugins/org.eclipse.elk.graph.json.text/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 37]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json.text/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.json/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 38]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.json/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.text.ide/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 39]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.text.ide/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.text.ui/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 40]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.text.ui/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

File: plugins/org.eclipse.elk.graph.text/.settings/org.eclipse.xtend.core.Xtend.prefs [hunk 41]
```diff
@@ -0,0 +1,6 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.graph.text/.settings/org.eclipse.xtend.core.Xtend.prefs:1]: Should not change

**Review by soerendomroes — CHANGES_REQUESTED** [review]
I suggest adding tests for your file system tree layout and make sure that independent plugins do not change and the java version remains 1.8 for now.

Moreover, you have to accept the eclipsefdn/eca

**Comment by uruuru** [comment]
> @uruuru @le-cds Do we have a policy on naming conventions for new algorithms and how do we usually integrate them?

We haven't done this before. Personally I'd expect `org.eclipse.elk...` identifiers that are in line with the existing ones.
